src/Disks/ObjectStorages/MetadataStorageFromDisk.cpp

```diff
@@ -100,6 +101,44 @@ DiskObjectStorageMetadataPtr MetadataStorageFromDisk::readMetadata(const std::st
     return readMetadataUnlocked(path, lock);
 }
 
+void MetadataStorageFromDisk::checkAndFixMetadataHardLinkUnlocked(const std::string & path, std::unique_lock<std::shared_mutex> & lock) const
+{
+    auto metadata = readMetadataUnlocked(path, lock);
+    auto index_path = metadata->getIndexPath();
+
+    if (!index_path.empty() && index_path != path)
```

Comment:
Empty. Index here is a hash of first S3 object name. Metadata format can contain multiple object for one "file". In most cases one metadata file contains one object. For some rare cases (for "Log" table engine) it can contains several object names. But also file can contains zero objects. I did not met this case never, I think that this is impossible according to ClickHouse logic, but I'm not sure that metadata file always contains at least one object or will do it in future. So this code will work in this impossible case.
index_path is a path in "index" directory. Method checks if "work" file is hardlink of "index" file and fix if not. But if it called for "index" file - then it should do nothing. `index_path != path` checks that path is not "index" path.

Reply:
> I did not met this case never, I think that this is impossible according to ClickHouse logic

Let's throw `LOGICAL_ERROR` then

> But if it called for "index" file - then it should do nothing

"index files" are internal implementation detail, why can it be called for index files? Anyway, please add a comment